Mathlib/Data/Array/extractLemmas.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2024 Jiecheng Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiecheng Zhao
+-/
+import Std.Data.Array.Lemmas
+/-!
+# lemmas about Array.extract
+
+Some useful lemmas about Array.extract
+-/
+set_option autoImplicit true
+
+namespace Array
+
+
+@[simp]
+theorem extract_zero_length {a : Array α} :
+    a.extract i i = #[] := by
+  refine extract_empty_of_stop_le_start a ?h
+  exact Nat.le_refl i
+
+theorem size_extract_to_stop {a: Array α} : (e ≤ a.size) → (a.extract s e).size = e - s := by
+  intro h1
+  simp only [size_extract]
+  rw [Nat.min_eq_left h1]
+
+theorem extract_append_left {a b: Array α} {i j : Nat} (h: j ≤ a.size):
+    (a ++ b).extract i j = a.extract i j := by
+  apply ext
+  · have h' : j ≤ a.size + b.size := by
+      apply Nat.le_trans h
+      exact Nat.le_add_right (size a) (size b)
+    simp only [size_extract, size_append, h', Nat.min_eq_left, h]
+  · intro h1 h2 h3
+    simp only [get_extract]
+    rw [get_append_left]
+
+theorem extract_append_right {a b: Array α} {i j : Nat} (h: i ≥ a.size):
+    (a ++ b).extract i j = b.extract (i - a.size) (j - a.size) := by
+  by_cases h1 : i ≥ j
+  · rw [extract_empty_of_stop_le_start _ h1, extract_empty_of_stop_le_start]
+    exact Nat.sub_le_sub_right h1 (size a)
+  by_cases h2 : i ≥ (a ++ b).size
+  · have h3:extract (a ++ b) i j = #[] := extract_empty_of_size_le_start (a ++ b) h2
+    have h4:extract b (i - size a) (j - size a) = #[] := by
+      apply extract_empty_of_size_le_start
+      rw [size_append] at h2
+      exact (Nat.le_sub_iff_add_le' h).mpr h2
+    rw [h3, h4]
+  · simp only [ge_iff_le, Nat.not_le] at h1
+    simp only [size_append, ge_iff_le, Nat.not_le] at h2
+    apply ext
+    · simp only [size_extract]
+      by_cases h3: j ≤ a.size + b.size
+      · have h4: j - a.size ≤ b.size := Nat.sub_le_iff_le_add'.mpr h3
+        simp only [size_append, h3, Nat.min_eq_left, h4]
+        rw [Nat.sub_sub, Nat.add_comm, Nat.sub_add_cancel h]
+      · have h3': j ≥ a.size + b.size := Nat.le_of_not_ge h3
+        have h4:  j - a.size ≥ b.size := Nat.le_sub_of_add_le' h3'
+        simp only [size_append, h3', Nat.min_eq_right, h4]
+        apply @Nat.add_right_cancel _ (i - a.size)
+        have h5 : i ≤ a.size + b.size := Nat.le_of_succ_le h2
+        have h6 : i - a.size ≤ b.size := Nat.sub_le_iff_le_add'.mpr h5
+        rw [← Nat.add_sub_assoc h, Nat.sub_add_cancel h5, Nat.add_comm,
+          Nat.add_sub_cancel, Nat.sub_add_cancel h6]
```

Comment:
Does `omega` end the proof earlier? Don't write out tedious `Nat` calculations when `omega` suffices.